### DISCUSSION TITLE
SBAT Handling in fwupd & LVFS

### DISCUSSION THREAD

**Discussion by superm1** [fwupd/fwupd#2954] [Ideas]
The [SBAT specification](https://github.com/rhboot/shim/blob/main/SBAT.example.md) explains exactly how SBAT is crafted and used to prevent having to fill a giant revocation database.

However, it doesn't discuss the details of how to distribute the SBAT variable signed by the UEFI CA, nor which application will be responsible for verifying it's safe to apply.

As such I had a proposal for fwupd to own this task, similar to how dbx updating is handled.

# Proposal

* No variable = "version 0"
* UEFI CA version number = sum(all numbers in SBAT).  That should mean any time there was effectively a revocation it goes up by 1 (or possibly more).
* To check it's safe to apply, verify that all binaries in $(ESP)\efi\$(VENDOR_PATH) have `.SBAT` region
* Check sbat region in all binaries, and compare against UEFI CA - abort if `component_generation` doesn't match and show error to user.
* If all looks good, load the authenticated variable.
* profit

# Distribution
If that flow works well and is agreeable, then LVFS can/should be positioned as the way to roll out new SBAT variables for future vulnerability disclosures.

**Comment by aburmash** [comment]
Overall your proposal looks good.
Few small comments:
1) Per current shim upstream last pull request, it is planned that variable will be enrolled by shim, in case there is no variable already set up.
So what is the " no variable" state per your proposal, no variable, or shim dummy variable state ?
https://github.com/rhboot/shim/pull/302

2) I feel that a better calculation may be needed for #2. Theoretically, there may be a case, when there is a new SBAT update, and sum of all numbers go down. So for example if SBAT variable included:
grub,1
grub.oracle,2 
grub.debian,2
// oracle and debian grubs contained a bug, not present in grub upstream or other vendor's grubs, and they had to be revoked by SBAT. We can't bump grub,1, because other vendors do not need to respin their grubs, only affected vendors need to.

But in some future, there may be a new generic grub2 update issued, all vendors migrate to grub,2 and vendor-specific entries may become redundant.

**Comment by superm1** [comment]
I wasn't aware of that PR.. I would ask why it needs to be bootservices though - isn't that going to mean it can only update through boot services too? In which case we would need to ferry the updated variable somewhere temporary like the ESP for "shim" to pick up and load the new one.

Regarding the count, maybe adding a weight to each layer is necessary then.

**Comment by superm1** [comment]
Here's a specific weighting proposal for version calculation.
1. Determine all of the top level prefixes in `SBAT` (IE `grub`, `fwupd`).
2. Multiply all top level prefixes by 100
3. `Version = sbat version + (sum of top level prefixes*100) + sum of second level prefixes`

Default version if that PR is merged:
```
sbat,1
```
1

So the example scenarios from spec:
```
sbat,1
shim,1
grub,1
grub.fedora,2
```
203

```
sbat,1
shim,1
grub,2
grub.fedora,2
```
303

```
sbat,1
shim,1
grub,3
grub.fedora,2
```
403

And a few others:
```
sbat,1
shim,3
grub,1
grub.fedora,2
grub.rhel,2
grub.debian,2
```
407

```
sbat,1
shim,3
grub,2
grub.fedora,2
grub.rhel,2
grub.debian,2
```
507

**Reply by hughsie** [reply]
I must admit I imagined it as `sbat_ver.sum(something).sum(something.something)` i.e. a `triplet` -- does that make more sense?

**Reply by superm1** [reply]
I think that's actually pretty similar to what I proposed (you're adding weight by keeping things a in certain place in the triplet).
It's certain to be more easy for a user to understand.
How about:

```
X.Y.Z
X:  sbat version
Y: sum(other top level objects)
Z: sum(all other objects)
```
